pkg/runner/mounting_runner.go

```diff
@@ -0,0 +1,105 @@
+package runner
+
+import (
+	"context"
+	"os"
+	"strings"
+
+	"github.com/buildbarn/bb-remote-execution/pkg/proto/configuration/bb_runner"
+	runner_pb "github.com/buildbarn/bb-remote-execution/pkg/proto/runner"
+	"github.com/buildbarn/bb-storage/pkg/filesystem"
+	"github.com/buildbarn/bb-storage/pkg/filesystem/path"
+	"github.com/buildbarn/bb-storage/pkg/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type mountingRunner struct {
+	base        runner_pb.RunnerServer
+	buildDirectory filesystem.Directory
+	mount *bb_runner.MountInfo
+}
+
+// NewMountingRunner is a decorator for Runner
+// that mounts `mount` before running a build action.
+//
+// This decorator can be used for chroot runners
+// that must mount special filesystems into the input root.
+func NewMountingRunner(base runner_pb.RunnerServer, buildDirectory filesystem.Directory, mount *bb_runner.MountInfo) runner_pb.RunnerServer {
+	return &mountingRunner{
+		buildDirectory: buildDirectory,
+		mount: mount,
+		base: base,
+	}
+}
+
+func open(buildDirectory filesystem.Directory, inputRootDirectory string) (filesystem.DirectoryCloser, error) {
+	// TODO(nils): How to best open the input root `Directory`
+	//             It must be a `localDirectory`, but outer directories can be `lazyDirectory`.
+	// NB: We iterate over Directory, as the initial buildDirectory is not a DirectoryCloser.
+	iterator := buildDirectory
+	var dir filesystem.DirectoryCloser
+
+	parts := strings.Split(inputRootDirectory, "/")
+	if len(parts) == 0 {
+		return nil, status.Error(codes.FailedPrecondition, "Unexpected empty inputRootDirectory")
+	}
+	for i, segment := range parts {
+		var err error
+		component := path.MustNewComponent(segment)
+		dir, err = iterator.EnterDirectory(component)
+		if i < len(parts) - 1 {
+			defer dir.Close()
+		}
+
+		iterator = dir.(filesystem.Directory)
+		if err != nil {
+			return nil, status.Errorf(codes.FailedPrecondition, "Could not enter directory component '%s' in '%#v'", segment, buildDirectory)
+		}
+	}
+	return dir, nil
+}
+
+func (r *mountingRunner) Run(ctx context.Context, request *runner_pb.RunRequest) (response *runner_pb.RunResponse, err error) {
+	root, err := open(r.buildDirectory, request.InputRootDirectory)
+	if err != nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "Could not enter input root directory '%s'", request.InputRootDirectory)
+	}
+	defer root.Close()
+
+	mountpoint := path.MustNewComponent(r.mount.Mountpoint)
+	err = root.Mkdir(mountpoint, 0o555)
+	if err != nil {
+		if ! os.IsExist(err) {
+			return nil, util.StatusWrapf(err, "Failed to create mount point: '%#v' in the input root", r.mount.Mountpoint)
```

Comment:
There is no need to put quotes around strings if you're using `%#v`. It already does that for you.